### PR TITLE
Implement subtitles and CC functionality for videos

### LINKED ISSUE #7479: Add Closed Captions & Subtitles to Video-Element
### Description

In order to move a little further towards accessibility in the future, it would be great if the video elements could also be provided with corresponding subtitles (e.g. as VTT).

See also https://developer.mozilla.org/en-US/docs/Web/Media/Audio_and_video_delivery/Adding_captions_and_subtitles_to_HTML5_video

### PR DESCRIPTION
**Pull request by zoglo**
### Description

Implements subtitles and closed caption support for videos (see: #7479)

Currently you can only choose between closed captions or subtitles, however both are almost treated the same.

_____

### Configuration

You can now enable subtitles/captions (and the other kinds) for video & audio (audio apparently allows them as well)

If you are adding 2 files in that order
- subtitles_de.vtt
- subtitles_en.vtt 

The first is chosen as the default.

Files of type `.vtt` will have 2 new fields that have to be setup:
`language` (select | mandatory)
`subtitle type (one out of the few options)

The `label` being mandatory is the `title` within the metadata

### CODE REVIEW

**Review by leofeyer — CHANGES_REQUESTED** [review]
Although it works, the current implementation is very error-prone.



- There could be more files than labels.
- The order of files could be different from the order of labels.
- There could be labels for files that don‘t exist.

The best solution I can think of is to use the file metadata instead. There is a `title` field already that can be used for the label, so all we have to do is to find a way to store the locale with the file.

**Review by fritzmg — CHANGES_REQUESTED** [review]
> * There could be more files than labels.
> * The order of files could be different from the order of labels.
> * There could be labels for files that don‘t exist.

I think all of this is acceptible.

* If there are more files than labels, then the additional files simply do not get the respective HTML attributes (unless they are mandatory - then we just do not output the files).
* If there are more labels than files, then the additional lables are simply ignored.
* If the order does not match up it's a user error. Not ideal, but imho a good enough compromise until we have a solution akin to `mvo/contao-group-widget` in the core. As discussed such a feature might get easer to implement in the future once we have introduced proper templates to our DC.

> There is a `title` field already that can be used for the label, so all we have to do is to find a way to store the locale with the file.

The locale is already stored in the meta wizard :). But storing it in the meta wizard is not ideal as that would require you to have a website root per language. The languages you provide for one video aren't necessarily the same languages that you provide on your website.

The **Subtitles kind** option does not make sense though. The VTT files can be a mix of both - this needs to be stored per file as well.

**Comment by leofeyer** [comment]
> I think all of this is acceptible.

It might be acceptable but its so easy to improve the UX here that it would be wrong not to do it 😄 

> The **Subtitles kind** option does not make sense though. The VTT files can be a mix of both - this needs to be stored per file as well.

That‘s just one more reason to enhance `tl_files`. A listener could add input fields for `tl_files.subtitlesLanguage` and `tl_files.subtitlesType` if the filetype is VTT.

**Comment by fritzmg** [comment]
> That‘s just one more reason to enhance `tl_files`. A listener could add input fields for `tl_files.subtitlesLanguage` and `tl_files.subtitlesType` if the filetype is VTT.

Sounds good to me 👍

**Comment by zoglo** [comment]
> That‘s just one more reason to enhance tl_files. A listener could add input fields for tl_files.subtitlesLanguage and tl_files.subtitlesType if the filetype is VTT.

That one was missing :(

> The Subtitles kind option does not make sense though. The VTT files can be a mix of both - this needs to be stored per file as well.

Till there's a proper widget to handle this, it's the known limitation to have a kind.

What we'd need right now would be a:
```File - Select (type) - Select (language) - Label```- Multicolumn-Widget
or the metadata :)

**Comment by leofeyer** [comment]
This is all we need IMHO:

**Comment by zoglo** [comment]
> This is all we need IMHO

We'd still need to check if label (title) or language are set because they'd be invalid without.

There is currently 4 subtitle kinds as well, the rest are `chapters` and `metadata` (metadata being not visible to users).
It also doesn't need a subtitle at all since it would fallback to `subtitles` if we leave it as a select (I had it as radio because someone said it'd be best practice for selections like that :) )

There's also the default flag but that should be the topmost added file imo.

**Comment by leofeyer** [comment]
> We'd still need to check if label (title) or language are set because they'd be invalid without.

We can show a warning in the player element if one of the subtitles files has incomplete metadata.

> It also doesn't need a subtitle at all since it would fallback to `subtitles`

Does this mean that the `kind` attribute is optional? And when it is not present, the browser assumes `subtitles`?

> There's also the default flag but that should be the topmost added file imo.

Yes 👍

**Comment by zoglo** [comment]
> We can show a warning in the player element if one of the subtitles files has incomplete metadata.

I agree

> Does this mean that the `kind` attribute is optional? And when it is not present, the browser assumes `subtitles`?

Yes

__

@fritzmg 
As discussed prior with @leofeyer , we do need kinds of subtitles since there is 4 mentioned ones in the moz docs but actually even 5 existing ones: https://html.spec.whatwg.org/multipage/media.html#the-track-element

And `captions` aka Closed Captions would be the most suitable for accessibility as well.

| Keyword      | State           | Brief description                                                                                                                                                                                                                     |
|--------------|-----------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `subtitles` | Subtitles | Transcription or translation of the dialogue, suitable for when the sound is available but not understood (e.g. because the user does not understand the language of the [media resource](https://html.spec.whatwg.org/multipage/media.html#media-resource)'s audio track). Overlaid on the video. |
| `captions` | Captions | Transcription or translation of the dialogue, sound effects, relevant musical cues, and other relevant audio information, suitable for when sound is unavailable or not clearly audible (e.g. because it is muted, drowned-out by ambient noise, or because the user is deaf). Overlaid on the video; labeled as appropriate for the hard-of-hearing. |
| `descriptions` | Descriptions | Textual descriptions of the video component of the [media resource](https://html.spec.whatwg.org/multipage/media.html#media-resource), intended for audio synthesis when the visual component is obscured, unavailable, or not usable (e.g. because the user is interacting with the application without a screen while driving, or because the user is blind). Synthesized as audio. |
| `chapters` | Chapters metadata | Tracks intended for use from script. Not displayed by the user agent.                                                                                                                                                                  |
| `metadata` | Metadata | Tracks intended for use from script. Not displayed by the user agent.                                                                                                                                                                                  |

It is also allowed to have multiple tracks and kinds such as in the example that is listed on the spec
```html
<video src="brave.webm">
    <track kind=subtitles src=brave.en.vtt srclang=en label="English">
    <track kind=captions src=brave.en.hoh.vtt srclang=en label="English for the Hard of Hearing">
    <track kind=subtitles src=brave.fr.vtt srclang=fr lang=fr label="Français">
    <track kind=subtitles src=brave.de.vtt srclang=de lang=de label="Deutsch">
</video>
```

**Comment by fritzmg** [comment]
> As discussed prior with @leofeyer , we do need kinds of subtitles since there is 4 mentioned ones in the moz docs but actually even 5 existing ones: https://html.spec.whatwg.org/multipage/media.html#the-track-element

Yes - but that was not my point. My point was it must not be a setting _per content element_. It must be a setting _per file_.

**Comment by zoglo** [comment]
> > As discussed prior with @leofeyer , we do need kinds of subtitles since there is 4 mentioned ones in the moz docs but actually even 5 existing ones: https://html.spec.whatwg.org/multipage/media.html#the-track-element
> 
> Yes - but that was not my point. My point was it must not be a setting _per content element_. It must be setting _per file_.

I'm on it already ;)

File: core-bundle/src/EventListener/DbafsMetadataSubscriber.php [hunk 1]
```diff
@@ -57,6 +57,14 @@ public function enhanceMetadata(RetrieveDbafsMetadataEvent $event): void
             $event->set('importantPart', $importantPart);
         }
 
+        // Add subtitle information
+        if (!empty($language = $row['subtitlesLanguage'] ?? null)) {
```
Comment (m-vo) [inline core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: Should we add a `Subtitle` class or sth alike for better DX (like we did for the other things)?

Comment (m-vo) [reply core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: This is how it could look:
```php
$subtitle = $file->getExtraMetadata()['subtitle']; // Class: Subtitle

$subtitle->getType();           // string-backed enum: SubtitleType
$subtitle->getSourceLanguage(); // string
$subtitle->getTitleLanguage();  // string

// Label:
$file->getExtraMetadata()['metadata']->getDefault()->getTitle();
```

I reckon having the types be an enum and using it in `subtitlesType.options` would be a good thing anyways.

Comment (zoglo) [reply core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: Added in:
2b4e15d20008b07f11c617605342721a9c19ccc2 5dc65b68f4420eadf65af682fe3f1838e0caae93 94ecb9e18ca738c2181f045be2664b7990fd0a35

File: core-bundle/src/File/Subtitle.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        private readonly string|null $type,
+    ) {
+    }
+
+    public function getSourceLanguage(): string
+    {
+        return $this->sourceLanguage;
+    }
+
+    public function getType(): string
```
Comment (m-vo) [inline core-bundle/src/File/Subtitle.php]: This needs to return the enum or `null`, not a string.

File: core-bundle/src/File/Subtitle.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+namespace Contao\CoreBundle\File;
+
+class Subtitle
+{
+    public function __construct(
+        private readonly string|null $sourceLanguage,
+        private readonly string|null $type,
+    ) {
```
Comment (m-vo) [inline core-bundle/src/File/Subtitle.php]: The constructor values should not include null (see getters).

File: core-bundle/src/File/SubtitleType.php [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\File;
+
+enum SubtitleType
```
Comment (m-vo) [inline core-bundle/src/File/SubtitleType.php]: These need to be string backed in order to be used as value for the database. (The names need to stay independent of the data.)

File: core-bundle/src/EventListener/DbafsMetadataSubscriber.php [hunk 5]
```diff
@@ -57,6 +58,12 @@ public function enhanceMetadata(RetrieveDbafsMetadataEvent $event): void
...
+            $subtitle = new Subtitle($language, $row['subtitlesType'] ?? null);
+            $event->set('subtitles', $subtitle);
```
Comment (m-vo) [inline core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: Singular, maybe?

**Review by m-vo — CHANGES_REQUESTED** [review]

File: core-bundle/src/EventListener/DbafsMetadataSubscriber.php (hunk 5 above)
Comment (leofeyer) [reply core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: No, that would be just one subtitle then. A .vtt file typically contains several entries, therefore "subtitles language" and "subtitles type" is correct.

Comment (m-vo) [reply core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: But the object currently only contains one of each. Is this supposed to be an array here, then?

Comment (fritzmg) [reply core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: @m-vo I think what @leofeyer means is that in a video you don't have one "subtitle" - it's always called "subtitles" (because it's many words, not just one word 🙃). Thus may be the class and enum should also be renamed to `Subtitles` and `SubtitlesType`.

Comment (m-vo) [reply core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: I see. Here, the metadata is not for a "video" but for a single file, though. Having a plural key there would IMHO suggest a collection of elements: `item.extraMetadata.subtitle.type` vs `item.extraMetadata.subtitles.type`. 

Alternatively we could go with `SubtitleAttributes`.

Comment (fritzmg) [reply core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: No, the point is the word "subtitle" is wrong in this context.

A "subtitle" is a secondary title to another heading. But in the context of videos it's always "subtitles", never _a_ subtitle.

(See also https://english.stackexchange.com/questions/526683/a-subtitle-vs-subtitles for instance)

Comment (m-vo) [reply core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: After rereading your comments, I think I'm getting your point: You could argue that a french `subtitle` is a single line of the french `subtitles` for instance.

Ok, fine with me. But let's have the class and property please be in line.

Comment (m-vo) [reply core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: > A "subtitle" is a secondary title to another heading. But in the context of videos it's always "subtitles", never a subtitle.

According to wiktionary, it's the singluar form for both: https://en.wiktionary.org/wiki/subtitle

Comment (m-vo) [reply core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: Can we agree on using `subtitles` as the key and `Subtitles` as the class name and `SubtitlesType` as the enum for consistency?

Comment (fritzmg) [reply core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: > According to wiktionary, it's the singluar form for both: https://en.wiktionary.org/wiki/subtitle

That's just how wiktionary is structured. But in the context of video you never say "subtitle", it's always "subtitles".

Comment (zoglo) [reply core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: Done in 79379327d064375187b5495040d1aba849018ec9

File: core-bundle/src/File/SubtitleType.php (hunk 4 above)
Comment (zoglo) [reply core-bundle/src/File/SubtitleType.php]: See 79379327d064375187b5495040d1aba849018ec9

File: core-bundle/src/File/Subtitle.php (hunk 3 above)
Comment (zoglo) [reply core-bundle/src/File/Subtitle.php]: See 79379327d064375187b5495040d1aba849018ec9

File: core-bundle/src/File/Subtitle.php (hunk 2 above)
Comment (zoglo) [reply core-bundle/src/File/Subtitle.php]: See 79379327d064375187b5495040d1aba849018ec9

**Comment by zoglo** [comment]
Expected PHP 8.1 check is unrelated and most likely due to raising min PHP-Version to PHP 8.2

File: core-bundle/src/Controller/ContentElement/PlayerController.php [hunk 6]
```diff
@@ -59,9 +59,16 @@ protected function getResponse(FragmentTemplate $template, ContentModel $model,
             return new Response();
         }
 
+        $subtitlesFiles = [];
+
+        if ($model->addSubtitles && $filesystemItems->first()?->isVideo()) {
+            $subtitlesItems = FilesystemUtil::listContentsFromSerialized($this->filesStorage, $model->subtitlesSRC ?: '');
+            $subtitlesFiles = $this->getSourceFiles($subtitlesItems);
+        }
+
```
Comment (m-vo) [inline core-bundle/src/Controller/ContentElement/PlayerController.php]: IMHO it's cleaner to you move this bit to `buildVideoFigureData()`. This would also remove the double `isVideo()` check.

**Comment by asaage** [comment]
Maybe you should refactor and name this thing "Text-Track(s)" instead of subtitle?
The html-element is ```<track>``` the file-extension is v**tt**. It can be subtitles, captions, a combination or something else.

File: core-bundle/src/Controller/ContentElement/PlayerController.php [hunk 7]
```diff
@@ -113,11 +121,42 @@ function (FilesystemItem $item) use (&$captions, $range): HtmlAttributes {
...
         );
 
+        $tracks = [];
+
+        $setDefault = false;
+
+        if ([] !== $subtitlesFiles) {
+            foreach ($subtitlesFiles as $file) {
+                $subtitles = $file->getExtraMetadata()['subtitles'] ?? null;
+                $label = ($file->getExtraMetadata()['metadata'] ?? null)?->getDefault()?->getTitle();
```
Comment (m-vo) [inline core-bundle/src/Controller/ContentElement/PlayerController.php]:
```suggestion
                $subtitles = $file->getExtraMetadata()->getSubtitles();
                $label = $file->getExtraMetadata()->getLocalized()?->getDefault()?->getTitle();
```
This needs #7626 to be merged upstream first. Then the `getSubtitles()` method can be added analogous to these: https://github.com/contao/contao/blob/ee27c852e2ea8dc0907e02e1e99ea346a80d57f4/core-bundle/src/Filesystem/ExtraMetadata.php#L56-L76

File: core-bundle/src/EventListener/DataContainer/SubtitlesFieldsListener.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        }
+
+        return $palette;
+    }
+
+    #[AsCallback(table: 'tl_files', target: 'fields.subtitlesType.options')]
+    public function subtitlesTypeOptions(): array
+    {
+        return array_map(static fn ($case) => $case->name, SubtitlesType::cases());
+    }
```
Comment (m-vo) [inline core-bundle/src/EventListener/DataContainer/SubtitlesFieldsListener.php]: I would be fine with adding this one line inline to the DCA file as well - similar to what you did for getting the locales. 

IIRC we discussed that options could allow enums as a first class citizen in the future (so that you could also just specify the enum class and the driver would then extract the cases). I reckon this would make it easier to refactor, then. 

@leofeyer Wdyt?

File: core-bundle/src/EventListener/DbafsMetadataSubscriber.php [hunk 9]
```diff
@@ -57,6 +59,12 @@ public function enhanceMetadata(RetrieveDbafsMetadataEvent $event): void
             $event->set('importantPart', $importantPart);
         }
 
+        // Add subtitles information
+        if (!empty($language = $row['subtitlesLanguage'] ?? null)) {
+            $subtitles = new Subtitles($language, $row['subtitlesType'] ? SubtitlesType::from($row['subtitlesType']) : null);
```
Comment (m-vo) [inline core-bundle/src/EventListener/DbafsMetadataSubscriber.php]: You can use [BackedEnum::tryFrom](https://www.php.net/manual/en/backedenum.tryfrom.php) here.

**Review by m-vo — CHANGES_REQUESTED** [review]
We're getting there. :heart:

File: core-bundle/src/EventListener/DataContainer/SubtitlesFieldsListener.php (hunk 8 above)
Comment (zoglo) [reply core-bundle/src/EventListener/DataContainer/SubtitlesFieldsListener.php]: > I would be fine with adding this one line inline to the DCA file as well

Was thinking the same once I coded this, just waiting for more opinions here

**Comment by m-vo** [comment]
I think @asaage has a point there. As briefly discussed with @zoglo on Slack, we should probably rethink the naming, because _now_ we've got types in our enumeration, that do not have anything to do with subtitles. 

My suggestion would be to…
* name the container class `TextTrack` with a key `textTrack` and a helper method `getTextTrack() : TextTrack|null`
* name the enum `TextTrackType`
… and keep everything else the same.

Then, at least from a technical standpoint, there would be no difference if the file provided a video overlay with captions, subtitles or descriptions.

File: core-bundle/src/Controller/ContentElement/PlayerController.php (hunk 7 above)
Comment (zoglo) [reply core-bundle/src/Controller/ContentElement/PlayerController.php]: Should somehow add a ToDo for that one since this would be unrelated to this PR but this looks way cleaner <3

Comment (m-vo) [reply core-bundle/src/Controller/ContentElement/PlayerController.php]: IMHO we should merge #7626 first and then adjust this PR. It cannot be done the other way round, because #7626 is targeting 5.3. Alternatively, we can merge this PR as is and create a followup PR. 

@leofeyer Any preferences?

File: core-bundle/src/Controller/ContentElement/PlayerController.php [hunk 10]
```diff
@@ -113,11 +113,48 @@ function (FilesystemItem $item) use (&$captions, $range): HtmlAttributes {
...
+                // Set the first file as the default track
+                if (!$setDefault) {
+                    $trackAttributes->set('default');
+                    $setDefault = true;
+                }
+
+                $tracks[] = $trackAttributes;
+            }
+        }
+
```
Comment (m-vo) [inline core-bundle/src/Controller/ContentElement/PlayerController.php]: I would clean up things in this section as follows, most importantly we should not reuse `getSourceFiles()` as it also sorts things and stuff.

```suggestion
        $tracks = [];

        if ($model->addTextTracks) {
            $trackItems = FilesystemUtil::listContentsFromSerialized($this->filesStorage, $model->textTrackSRC ?: '');

            foreach ($trackItems as $trackItem) {
                if (!$publicUri = $this->filesStorage->generatePublicUri($trackItem->getPath())) {
                    continue;
                }

                $extraMetadata = $trackItem->getExtraMetadata();

                /** @todo change to: … = $extraMetadata->getTextTrack(); */
                if (null === ($textTrack = $extraMetadata['textTrack'] ?? null)) {
                    continue;
                }

                /** @todo change to: … = $extraMetadata->getLocalized()?->getDefault()?->getTitle(); */
                if (null === ($label = ($extraMetadata['metadata'] ?? null)?->getDefault()?->getTitle())) {
                    continue;
                }

                $tracks[] = (new HtmlAttributes())
                    ->setIfExists('kind', $textTrack->getType()?->value)
                    ->set('label', $label)
                    ->set('srclang', $textTrack->getSourceLanguage())
                    ->set('src', $publicUri)
                ;
            }

            // Set the first file as the default track
            ($tracks[0] ?? null)?->set('default');
        }

```

**Review by m-vo — CHANGES_REQUESTED** [review]

**Comment by zoglo** [comment]
This is ready for another review (and hopefully ready to merge)

There is one last conversation that could be ignored or not / could be implemented in a follow-up PR.

Just an info for @leofeyer:

d734ccb0acbd51f9a32b9d51fcd9372296619d82 added an error message when the files do not have the proper metadata set and will add an error message:

image

File: core-bundle/src/Controller/ContentElement/PlayerController.php (hunk 7 above)
Comment (leofeyer) [reply core-bundle/src/Controller/ContentElement/PlayerController.php]: #7626 has been merged 👍

Comment (zoglo) [reply core-bundle/src/Controller/ContentElement/PlayerController.php]: @m-vo @leofeyer done in 621127de1aa16f3ddddd51cb8010f9cb81bdf000 :)

File: core-bundle/src/EventListener/DataContainer/TextTrackFieldsListener.php [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+use Contao\CoreBundle\DependencyInjection\Attribute\AsCallback;
+use Contao\DataContainer;
+
+class TextTrackFieldsListener
+{
+    #[AsCallback(table: 'tl_files', target: 'config.onpalette')]
+    public function addTextTrackFields(string $palette, DataContainer $dc): string
+    {
+        // $dc->id is the file name in this case
+        if (str_ends_with($dc->id, '.vtt')) {
```
Comment (m-vo) [inline core-bundle/src/EventListener/DataContainer/TextTrackFieldsListener.php]: Nitpick: This comparison should probably use `Path::getExtension()` with `forceLowercase` set to `true`. Otherwise for example a `foo.VTT` file would not be recognized.

**Review by m-vo — DISMISSED** [review]

**Comment by m-vo** [comment]
Remarks:

*  I did not review things in the `TrackTitleSourceListener` class, because I don't know the "correct way" to do add messages. This class may also deserve a unit test :see_no_evil:.

File: core-bundle/src/EventListener/DataContainer/TextTrackFieldsListener.php (hunk 11 above)
Comment (zoglo) [reply core-bundle/src/EventListener/DataContainer/TextTrackFieldsListener.php]: See 10fe148937804ff1b5068b57d2ffd5777eac707b

Comment (m-vo) [reply core-bundle/src/EventListener/DataContainer/TextTrackFieldsListener.php]: then obviously without the `str_ends_with` :smile:

Comment (zoglo) [reply core-bundle/src/EventListener/DataContainer/TextTrackFieldsListener.php]: OOOH :D

**Review by m-vo — DISMISSED** [review]

**Comment by aschempp** [comment]
I always wonder why we have a checkbox that enables a mandatory selector field, instead of just having a non-mandatory file picker? It seems kind of common in Contao, but not sure what is better? It would save us a DB field and a click to the user. 🤷‍♂️

**Comment by zoglo** [comment]
> I always wonder why we have a checkbox that enables a mandatory selector field, instead of just having a non-mandatory file picker? It seems kind of common in Contao, but not sure what is better? It would save us a DB field and a click to the user. 🤷‍♂️

Because new fields (plural) are added when using the checkbox (has been the case in the first state of the PR).
I like your idea, we are only adding one field (after moving everything else into the file metadata) so it's actually not needed.

**Comment by zoglo** [comment]
@aschempp see bb0209589b324ec5b57b10a7fa427cc8e49bde50

Since it's null (even after saving), this works. Does not need to be mandatory at all, the rest is being handled within the listener and PlayerController (There was also no need to adjust the tests since resetting the error message would need to happen if we clear all files and there has been an invalid one).

image

@leofeyer This PR is done, feel free to review (and also adjust the screenshot on contao.org if you approve since the checkbox is gone now :) (I adjusted the translation to your idea)

File: core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        if ([] !== $invalid) {
+            $message->addError(\sprintf($GLOBALS['TL_LANG']['ERR']['textTrackMetadataMissing'], implode(', ', $invalid)));
+        } elseif ($message->hasError()) {
+            $session = $this->requestStack->getCurrentRequest()?->getSession();
+
+            // Reset the error if it exists
+            if ($session instanceof FlashBagAwareSessionInterface) {
+                $session->getFlashBag()->get('contao.BE.error');
+            }
+        }
```
Comment (fritzmg) [inline core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php]: Why is this needed? Only adding the message should be fine.

Comment (zoglo) [reply core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php]: It's for removing the previous message when adjusting the files. Without it, it would still stay after saving again.

Comment (fritzmg) [reply core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php]: Hm, that should not be the case. Also this would apply to _any_ error, not just the one you added. Also why does simply calling `$session->getFlashBag()->get('contao.BE.error');` reset it?

Comment (zoglo) [reply core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php]: https://github.com/contao/contao/blob/2c3f94f08022152055225fa1ba526f07c29ddaa9/core-bundle/contao/library/Contao/Message.php#L187

https://github.com/symfony/symfony/blob/b30694fe182926e36defd1d5c3266d949c0187d2/src/Symfony/Component/HttpFoundation/Session/Flash/FlashBagInterface.php#L50-#L55

```php
    /**
     * Gets and clears flash from the stack.
     *
     * @param array $default Default value if $type does not exist
     */
    public function get(string $type, array $default = []): array;
```

Comment (zoglo) [reply core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php]: ^ Basically because it sets an empty Array of messages to the flashbag so that is the intended way to use it.

Comment (fritzmg) [reply core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php]: Yes, but you should just use `$message->reset()` then. In any case, I would inspect why the flashbag is displayed again, this should not happen. Also I would make the following additional changes:

1. Remove `submitOnChange` from `tl_content.textTrackSRC`.
2. Remove `#[AsCallback(table: 'tl_content', target: 'fields.textTrackSRC.load')]` from `TrackTitleSourceListener`.

Comment (fritzmg) [reply core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php]: Ah, I see the problem. Our `Message::add()` lets the message display for the current request _and_ the next request through the flashbag. Meaning that the `load_callback` will add the message for the current request (which is correct) but will also show it for the next request - where the file might have already been fixed.

File: core-bundle/contao/dca/tl_content.php [hunk 13]
```diff
@@ -602,6 +602,12 @@
 			'eval'                    => array('includeBlankOption' => true, 'nospace'=>true, 'tl_class'=>'w50'),
 			'sql'                     => "varchar(8) COLLATE ascii_bin NOT NULL default ''"
 		),
+		'textTrackSRC' => array
+		(
+			'inputType'               => 'fileTree',
+			'eval'                    => array('multiple'=>true, 'fieldType'=>'checkbox', 'submitOnChange'=>true, 'filesOnly'=>true, 'extensions'=>'vtt', 'isSortable'=>true),
```
Comment (fritzmg) [inline core-bundle/contao/dca/tl_content.php]:
```suggestion
			'eval'                    => array('multiple'=>true, 'fieldType'=>'checkbox', 'filesOnly'=>true, 'extensions'=>'vtt', 'isSortable'=>true),
```

This does not seem to be necessary and is otherwise irritating, as `submitOnChange` on non-checkboxes cause a complete reresh.

**Review by fritzmg — CHANGES_REQUESTED** [review]

File: core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php (hunk 12 above)
Comment (zoglo) [reply core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php]: See 5af6ce9f2e5cd1365e2b2285fabcc418fb56ea92

Added a `ToDo` as mentioned in Slack due to the message being kept even within the next request

File: core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+        $message = $this->framework->getAdapter(Message::class);
+
+        if ([] !== $invalid) {
+            $message->addError(\sprintf($GLOBALS['TL_LANG']['ERR']['textTrackMetadataMissing'], implode(', ', $invalid)));
+        } elseif ($message->hasError()) {
+            // ToDo: Resets the message - There is currently no possible way to show messages
+            // for just the current request
```
Comment (fritzmg) [inline core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php]:
```suggestion
            // TODO: We need to reset all messages here as there is currently no way to show 
            // a message just for the current request.
```

**Review by fritzmg — DISMISSED** [review]

File: core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+                ) {
+                    $invalid[] = $fileSystemItem->getName();
+                }
+            }
+        }
+
+        $message = $this->framework->getAdapter(Message::class);
+
+        if ([] !== $invalid) {
+            $message->addError(\sprintf($GLOBALS['TL_LANG']['ERR']['textTrackMetadataMissing'], implode(', ', $invalid)));
```
Comment (leofeyer) [inline core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php]: We should inject the `translator` service instead of using the `$GLOBALS['TL_LANG']` array.

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+        } elseif ($message->hasError()) {
+            // TODO: We need to reset all messages here as there is currently no way to show
+            // a message just for the current request.
+            $message->reset();
```
Comment (leofeyer) [inline core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php]: I don‘t understand what problem exactly this solves?

Comment (fritzmg) [reply core-bundle/src/EventListener/DataContainer/TrackTitleSourceListener.php]: See the discussion [here](https://github.com/contao/contao/pull/7604#discussion_r1818746787) - currently our `Message` class cannot handle messages that should only be displayed for the _current_ request. `Message::add()` will always add it to the flash bag of the session, meaning it will be always also displayed for the _next_ request.

This is a workaround until we have another solution for this.

**Review by leofeyer — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @zoglo.
